### PR TITLE
raft: introduce/fix test_node_with_smaller_term_can_complete_election

### PR DESCRIPTION
**Pull request by july2993**
contirbute to https://github.com/pingcap/raft-rs/issues/21

### CODE REVIEW

File: tests/cases/test_raft.rs [hunk 1]
```diff
@@ -3462,6 +3478,67 @@ fn test_transfer_non_member() {
...
+    nt.recover();
+    nt.cut(2, 1);
+    nt.cut(2, 3);
+
+    // call for election
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    // do we have a leader?
+    assert!(nt.peers[&1].state == StateRole::Leader || nt.peers[&3].state == StateRole::Leader);
```
Comment (siddontang) [inline tests/cases/test_raft.rs]:
```
assert!(check, "no leader")
```

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

File: tests/cases/test_raft.rs [hunk 2]
```diff
@@ -1758,9 +1758,25 @@ fn test_recv_msg_request_vote_for_type(msg_type: MessageType) {
...
+        m.set_log_term(log_term);
+        // raft.Term is greater than or equal to raft.raftLog.lastTerm. In this
+        // test we're only testing MsgVote responses when the campaigning node
+        // has a different raft log compared to the recipient node.
+        // Additionally we're verifying behaviour when the recipient node has
+        // already given out its vote for its current term. We're not testing
+        // what the recipient node does when receiving a message with a
+        // different term number, so we simply initialize both term numbers to
+        // be the same.
+        let term = if sm.raft_log.last_term() > log_term {
```
Comment (BusyJay) [inline tests/cases/test_raft.rs]: You can use `cmp::max` here.

File: tests/cases/test_raft.rs [hunk 3]
```diff
@@ -1770,7 +1786,7 @@ fn test_recv_msg_request_vote_for_type(msg_type: MessageType) {
         if msgs[0].get_msg_type() != vote_resp_msg_type(msg_type) {
             panic!(
                 "#{}: m.type = {:?}, want {:?}",
-                i,
+                index,
```
Comment (BusyJay) [inline tests/cases/test_raft.rs]: I think it should be `j`.

**Pull request merged**